src/algebra/star/basic.lean

```diff
@@ -118,32 +120,45 @@ local attribute [instance] star_monoid_of_comm
 end
 
 /--
-A `*`-ring `R` is a (semi)ring with an involutive `star` operation which is additive
-which makes `R` with its multiplicative structure into a `*`-monoid
-(i.e. `star (r * s) = star s * star r`).
+A `*`-additive monoid `R` is an additive monoid with an involutive `star` operation which
+preserves addition.
 -/
-class star_ring (R : Type u) [semiring R] extends star_monoid R :=
+class star_add_monoid (R : Type u) [has_add R] extends has_involutive_star R :=
 (star_add : ∀ r s : R, star (r + s) = star r + star s)
 
-@[simp] lemma star_add [semiring R] [star_ring R] (r s : R) : star (r + s) = star r + star s :=
-star_ring.star_add r s
+export star_add_monoid (star_add)
+attribute [simp] star_add
+
+instance [has_add R] [star_add_monoid R] : star_add_monoid (Rᵒᵖ) :=
+{ star_add := λ x y, unop_injective (star_add x.unop y.unop) }
 
 /-- `star` as an `add_equiv` -/
 @[simps apply]
-def star_add_equiv [semiring R] [star_ring R] : R ≃+ R :=
+def star_add_equiv [has_add R] [star_add_monoid R] : R ≃+ R :=
 { to_fun := star,
   map_add' := star_add,
   ..(has_involutive_star.star_involutive.to_equiv star)}
 
 variables (R)
 
-@[simp] lemma star_zero [semiring R] [star_ring R] : star (0 : R) = 0 :=
+@[simp] lemma star_zero [add_zero_class R] [star_add_monoid R] : star (0 : R) = 0 :=
 (star_add_equiv : R ≃+ R).map_zero
 
 variables {R}
 
+/--
+A `*`-ring `R` is a (semi)ring with an involutive `star` operation which is additive
+which makes `R` with its multiplicative structure into a `*`-monoid
+(i.e. `star (r * s) = star s * star r`).
+-/
+class star_ring (R : Type u) [semiring R] extends star_monoid R :=
```

Comment:
Does it not work to simultaneously extend `star_monoid` and `star_add_monoid`?

Reply:
Nope, these are new-style classes which don't support that